File: minmax_charspeed.cpp
#include "minmax_charspeed.h"
#include "update_utilities.h"

void minmax_charspeed(Matrix<double> &ua, Matrix<double> &Sa)
{
	struct test_cases* pt_test = get_tests();
	
	int N2 = ua.rows();
	int M2 = ua.cols() - 2 * pt_test->gc;

	Matrix<double> *ptul = new Matrix<double>(N2, 1); Matrix<double> &ul = *ptul;
	Matrix<double> *ptur = new Matrix<double>(N2, 1); Matrix<double> &ur = *ptur;
	Matrix<double> *ptuh = new Matrix<double>(3, 1);  Matrix<double> &uh = *ptuh;
	Matrix<double> *ptSh = new Matrix<double>(2, M2 + 2 * pt_test->gc); Matrix<double> &Sh = *ptSh;
	
	for (int i = pt_test->gc + 1; i <= M2 + pt_test->gc; i++)
	{
		get_col(ul, ua, i);
		get_col(ur, ua, i + 1);
		lminmax_charspeed(ul, ur, uh);
		for (int j = 1; j <= 2; j++) Sa(j, i) = uh(j);
	}
	delete ptSh;
	delete ptuh;
	delete ptul;
	delete ptur;
}

File: minmax_charspeed.m
function Sa=minmax_charspeed(ua, Sa)

%int: M2, N2, i, j, gc
%vector: uah, uai, ul, ur, uh
%matrix: ua, Sa

global gc;

M2=size(ua, 2)-2*gc;
N2=size(ua, 1);

ul=zeros(N2,1);
ur=zeros(N2,1);
uh=zeros(3,1);

for i = gc+1:M2+gc
    for j = 1:N2
        ul(j)=ua(j, i);
        ur(j)=ua(j, i+1);
    end
    uh=lminmax_charspeed(ul, ur, uh);
    for j = 1:2
        Sa(j, i)=uh(j);
    end
end